### PR TITLE
11.0 fix css massmailing chm

### PR DESCRIPTION
**Pull request by Gorash**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Gorash** [comment]
@qsm-odoo tu peux regarder rapidement? ;)

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 1]
```diff
@@ -229,6 +229,9 @@ function classToStyle($editable) {
         } else {
             $target.attr('style', style);
         }
+        if (this.nodeName === 'TD' && !this.childNodes.length) {
+            this.innerHTML = '&nbsp;';
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: Typo in the commit message id -> if

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 2]
```diff
@@ -105,25 +105,42 @@ function getMatchedCSSRules(a) {
         delete style.display;
     }
 
-    _.each(['margin', 'padding'], function (p) {
-        if (style[p+'-top'] || style[p+'-right'] || style[p+'-bottom'] || style[p+'-left']) {
-            if (style[p+'-top'] === style[p+'-right'] && style[p+'-top'] === style[p+'-bottom'] && style[p+'-top'] === style[p+'-left']) {
+    _.each([['margin'], ['padding'], ['border', 'style']], function (attr) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: Commit message should explain this and inline comments should explain this new transcoding behavior

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 3]
```diff
@@ -105,25 +105,42 @@ function getMatchedCSSRules(a) {
...
+    if (style['border-bottom-left-radius']) {
+        style['border-radius'] = style['border-bottom-left-radius'];
+        delete style['border-bottom-left-radius'];
+        delete style['border-bottom-right-radius'];
+        delete style['border-top-left-radius'];
+        delete style['border-top-right-radius'];
+    }
+
+    _.each(_.keys(style), function (k) {
+        if (k.indexOf('border') !== -1 && style[k] === 'initial') {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: I am not sure about this one for example? I want moar doc :money_mouth_face:

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 4]
```diff
@@ -229,9 +246,30 @@ function classToStyle($editable) {
...
             $target.attr('style', style);
         }
+        // Apple Mail
         if (this.nodeName === 'TD' && !this.childNodes.length) {
             this.innerHTML = '&nbsp;';
         }
+
+        // Outlook
+        if (this.nodeName === 'A' && $target.hasClass('btn') && !$target.children().length) {
+            var $hack = $('<table class="o_outlook_hack"><tr><td></td></tr></table>');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: I am sure other than outlook have bugs too. I would name this "o_table_btn_hack" (so that we know it is a button too). But do as you wish.

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 5]
```diff
@@ -229,9 +246,30 @@ function classToStyle($editable) {
...
+            $hack.find('td')
+                .attr('height', $target.outerHeight())
+                .css('margin', $target.css('padding'))
+                .css('border-radius', $target.css('border-radius'))
+                .css('background-color', $target.css('background-color'));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]:
```javascript
.css({
    margin: ...,
    'border-radius': ...,
    ...
})
```

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 6]
```diff
@@ -229,9 +246,30 @@ function classToStyle($editable) {
...
+            $target.after($hack);
+            $target.appendTo($hack.find('td'));
+            var node = $hack[0].previousSibling;
+            if (node && node.nodeType === 3 && !node.textContent.match(/\S/)) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: `Node.TEXT_NODE`
(and again: why is this needed ?)

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 7]
```diff
@@ -242,6 +280,11 @@ function classToStyle($editable) {
  * @param {jQuery} $editable
  */
 function styleToClass($editable) {
+    // Outlook revert
+    $editable.find('table.o_outlook_hack').each(function () {
+        $(this).after($('a', this));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: Wut ? I do not understand that line. And don't you want to just extract the button from the table ?

File: addons/web_editor/static/src/js/editor/transcoder.js [hunk 8]
```diff
@@ -229,9 +246,30 @@ function classToStyle($editable) {
...
+        // Outlook
+        if (this.nodeName === 'A' && $target.hasClass('btn') && !$target.children().length) {
+            var $hack = $('<table class="o_outlook_hack"><tr><td></td></tr></table>');
+            $hack.find('td')
+                .attr('height', $target.outerHeight())
+                .css('margin', $target.css('padding'))
+                .css('border-radius', $target.css('border-radius'))
+                .css('background-color', $target.css('background-color'));
+            $target.after($hack);
+            $target.appendTo($hack.find('td'));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/transcoder.js]: So basically you are putting the button in a table cell with the same style of the button... is not this gonna duplicate the margin for smarter mail clients ?

Comment (Gorash) [reply addons/web_editor/static/src/js/editor/transcoder.js]: The rendering is the same in the tests, both are collapsing (and one is not interpreted by some browsers)

File: addons/web_editor/static/src/js/editor/transcoder.js (hunk 6 above)
Comment (Gorash) [reply addons/web_editor/static/src/js/editor/transcoder.js]: because, the are some time a white space unbreakable and this char add a line when is a table and is invisible when it's a link
